File: Leetcode/tree/590_N-aryTreePostorderTraversal/Solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    /*/ //recursive solution
    public List<Integer> postorder(Node root) {
        List<Integer> res = new LinkedList<>();
        preorderTraversal(res, root);
        return res;
    }
        private void preorderTraversal(List<Integer> res, Node root){
            if(root==null) return;
            for(Node child: root.children){
                preorderTraversal(res, child);
            }
            res.add(root.val);
        }/*/
    // recursive solution2
    public List<Integer> postorder(Node root){
        List<Integer> res = new LinkedList<>();
        if(root == null)
            return res;
        for(Node child : root.children) {
            res.addAll(postorder(child));
        }
        res.add(root.val);
        return res;
    }
    /*/ //iterative solution
    public List<Integer> postorder(Node root){
        LinkedList<Integer> res = new LinkedList<>();
        if (root == null) return res;
        Stack<Node> stack = new Stack<>();
        stack.push(root);
        while (!stack.empty()) {
            Node temp = stack.pop();
            res.add(0, temp.val);
            for (int i = 0; i < temp.children.size(); i++)
                stack.push(temp.children.get(i));
        }
        return res;
    }/*/
}

File: Leetcode/tree/590_N-aryTreePostorderTraversal/Solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
public:
    // iterative
    vector<int> postorder(Node* root) {
        vector<int> res;
        if(root==nullptr) return res;
        stack<Node*> stk;
        stk.push(root);
        while(!stk.empty())
        {
            Node* temp=stk.top();
            stk.pop();
            for(auto child: temp->children) stk.push(child);
            res.push_back(temp->val);
        }
        reverse(res.begin(), res.end());
        return res;
    }
    // recursive
    /*/
    vector<int> postorder(Node* root) {
        vector<int> res;
        preorder_trav(&res, root);
        return res;
    }
    void preorder_trav(vector<int>* res, Node* root){
        if(root==nullptr)
            return;
        for(auto child: root->children){
            preorder_trav(res, child);
        }
        res->push_back(root->val);
    }/*/
};
